server/base/src/main/java/org/apache/accumulo/server/util/ZooZap.java

```diff
@@ -115,6 +128,71 @@ public void zap(SiteConfiguration siteConf, String... args) {
+          final String zkRoot = ZooUtil.getRoot(iid);
+          final String upgradePath = zkRoot + Constants.ZPREPARE_FOR_UPGRADE;
+
+          try {
+            if (zrw.exists(upgradePath)) {
+              if (!opts.forceUpgradePrep) {
+                throw new IllegalStateException(
+                    "'ZooZap -prepare-for-upgrade' must have already been run."
+                        + " To run again use the 'ZooZap -prepare-for-upgrade -force'");
+              } else {
+                zrw.delete(upgradePath);
+              }
+            }
+          } catch (KeeperException | InterruptedException e) {
+            throw new IllegalStateException("Error creating or checking for " + upgradePath
+                + " node in zookeeper: " + e.getMessage(), e);
+          }
+
+          log.info("Upgrade specified, validating that Manager is stopped");
+          if (context.getServerPaths().getManager(true) != null) {
+            throw new IllegalStateException(
+                "Manager is running, shut it down and retry this operation");
+          }
+
+          log.info("Checking for existing fate transactions");
+          try {
+            // Adapted from UpgradeCoordinator.abortIfFateTransactions
+            if (!zrw.getChildren(context.getZooKeeperRoot() + Constants.ZFATE).isEmpty()) {
+              throw new IllegalStateException("Cannot complete upgrade preparation"
+                  + " because FATE transactions exist. You can start a tserver, but"
+                  + " not the Manager, then use the shell to delete completed"
+                  + " transactions and fail pending or in-progress transactions."
+                  + " Once all of the FATE transactions have been removed you can"
+                  + " retry this operation.");
+            }
+          } catch (KeeperException | InterruptedException e) {
+            throw new IllegalStateException("Error checking for existing FATE transactions", e);
+          }
+
+          log.info("Creating {} node in zookeeper, servers will be prevented from"
+              + " starting while this node exists", upgradePath);
+          try {
+            zrw.putPersistentData(upgradePath, new byte[0], NodeExistsPolicy.SKIP);
+          } catch (KeeperException | InterruptedException e) {
+            throw new IllegalStateException("Error creating " + upgradePath
+                + " node in zookeeper. Check for any issues and retry.", e);
+          }
+          log.info("Instance {} prepared for upgrade. Server processes will not start while"
+              + " in this state. To undo this state and abort upgrade preparations delete"
+              + " the zookeeper node: {}", iid.canonical(), upgradePath);
+
+          log.info("Forcing removal of all server locks");
+          // modify the options to remove all locks
+          opts.zapCompactors = true;
+          opts.zapManager = true;
+          opts.zapScanServers = true;
+          opts.zapTservers = true;
```

Comment:
What about GC and Monitor?

Reply:
I think that should be a separate PR. ZooZap has never removed locks for these types of servers.

Reply:
Bumped into this in #5439. Had to add the other server types there to make pruning work.

Reply:
https://github.com/apache/accumulo/issues/5642 reminded me of my comment here; good to see this was fixed :+1: